aten/src/ATen/native/TensorAdvancedIndexing.cpp

```diff
@@ -470,23 +503,32 @@ Tensor & _index_put_impl_(Tensor & self, const torch::List<c10::optional<Tensor>
       "Please clone() the tensor before performing this operation. "
       "This also applies to advanced indexing e.g. tensor[indices] = tensor");
   }
+  if (!accumulate) {
+    auto masked_fill_dispatch = canDispatchToMaskedFill(self, indices, value);
+    if (std::get<0>(masked_fill_dispatch)) {
+      return self.masked_fill_(std::get<1>(masked_fill_dispatch), value.item());
+    }
+  }
+  auto value_ = value;
+  if (value.device() != self.device() && value.numel() == 1) {
```

Comment:
I think the appropriate test here (to preserve current behavior) is `is_wrapped_number`, 1-element tensors on the different device error out:
```
In [7]: a=torch.randn(4, device="cuda")

In [8]: b=torch.full((), 1)

In [9]: a[[0,1,2]]=b
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-9-16e5c5d62ea1> in <module>
----> 1 a[[0,1,2]]=b

RuntimeError: Index put requires the source and destination dtypes match, got Float for the destination and Long for the source.
```
I hope the way you are wrapping python scalars is correctly setting `is_wrapped_number` attribute.